patterns/interpreter.md

```diff
@@ -0,0 +1,272 @@
+# Interpreter
+
+## Description
+
+If a problem occurs very often and requires long and repetitive steps to solve it,
+then the problem instances might be expressed in a simple
+language and inerpreter object could solve it by
+interpreting the sentences written in this simple language.
+Basically, for any kind of problems we define a domain
+language, then define a grammar for this language and
+design interpreter solving problem instances.
+
+## Motivation
+
+Imagine that our work is translating simple mathematical expressions into
+[assembly language](https://en.wikipedia.org/wiki/Assembly_language)
+(more simple and low level programming language).
+For simplicity, our expressions consists of ten digits `0`,...,`9`,
+four operations `+, -, /, *` and a pair of parenthesis `(, )`.
+For example, expression `2 + 4` could be translated into
+
+```ignore
+mov eax, 2
+mov ebx, 4
+add eax, ebx
+```
+
+Our goal is to automate translating into assembly instructions
+using the Interpreter design pattern. In other words, we want simply
+provide the Interpreter with an expressions and get Assembly
+language output. For example
+
+```rust, ignore
+x.interpret("7+3*(2-1)", &output);
+```
+
+We don't claim that output assembly code is exactly what modern compilers generate,
+but it at least computes the expression correctly.
+It is simpy an example of using Interpreter pattern.
+
+## Solution 1
+
+The grammar for a set of expressions over `0,...,9, +,-,*,/,(,)` is
```

Comment:
Should we give a brief explanation about what a grammar is? And explain this example in particular?
For example why both `exp` and `term` are necessary.
Otherwise this chapter is readable only by people with a MSc degree in computer science 😅

Reply:
@MarcoIeni  Sure, I can explain, but I think we shouldn't go beyond simply mentioning what a CFG is and that `expr`, `terms...` and `factor` are the result of grammar transformation. In any case strong knowledge of CS is still needed :-D . I don't mind detailed explanation of CFG and related topics, but to what extend?

Reply:
Thank you for providing the small explanation, with an example, too!
I think it works great :)